agent/cache/watch.go

```diff
@@ -106,148 +106,192 @@ func (c *Cache) notifyBlockingQuery(ctx context.Context, r getOptions, correlati
 
 	for {
 		// Check context hasn't been canceled
-		if ctx.Err() != nil {
+		select {
+		case <-ctx.Done():
 			return
+		default:
 		}
 
-		// Blocking request
-		r.Info.MinIndex = index
-		res, meta, err := c.getWithIndex(ctx, r)
+		index, failures = c.notifyBlockingQueryOnce(ctx, r, correlationID, cb, index, failures)
+	}
+}
+func (c *Cache) notifyBlockingQueryOnce(
+	ctx context.Context,
+	r getOptions,
+	correlationID string,
+	cb Callback,
+	currIndex uint64,
+	currFailures uint,
+) (index uint64, failures uint) {
+	// Seed response data from last execution.
+	index = currIndex
+	failures = currFailures
 
-		// Check context hasn't been canceled
-		if ctx.Err() != nil {
-			return
-		}
+	// Ensure that each invocation terminates instead of stacking.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
Wouldn't this cancel the context as soon as it's out of scope at the end of `notifyBlockingQueryOnce`?
If so that would terminate the go routine inside fetch for the current iteration not the former one. Am I missing something here?

Reply:
So I moved the body of the cancelable portion into `notifyBlockingQueryOnce` so that I could `defer` the cancel and avoid trying to do that safely without defer in a loop.

If you look at `notifyBlockingQuery` it doesn't even sleep so it immediately re-enters `notifyBlockingQueryOnce` and creates a new one.

In a prior iteration of this I tried to do the "new one cancels the prior one" but 
a) it had some gross conditional cancellation
b) the linter had red flags go off at suspected escaped cancels because of (a)